crates/proof-of-sql/src/base/database/column_index_operation.rs

```diff
@@ -0,0 +1,157 @@
+use super::{slice_operation::apply_slice_to_indexes, Column, ColumnOperationResult, ColumnType};
+use crate::base::scalar::Scalar;
+use alloc::vec::Vec;
+use bumpalo::Bump;
+
+#[allow(dead_code)]
+pub trait IndexOp {
```

Comment:
I'm curious what other use cases this abstraction will have?

I see that implementing applying an index to a column is aided by using a generic function that does the same on a generic slice, but couldn't we just have a method on `Column` like `fn apply_index(self, indixes &[usize], alloc: &'a Bump) -> Column`?

I see that this abstraction allows you to essentially also apply some generic function to the column you're applying the index to in the process, but that potential is unused in this PR and could be done by applying a more basic abstraction for writing generic-over-column-type functions like after applying the index. Something like `trait GenericOverColumnTypeFn` that has a `fn call<T: PrimitiveColumnType>(value: T) -> T` and a method on `Column` like `fn generic_map(f: GenericOverColumnTypeFn, alloc: &'a Bump) -> Column`

Anyway, if this "writing a function for generic slices that takes indexes" has more use cases than just applying an index, or maybe applying an index and a function, I'd be interested to hear them.

Reply:
You are right that in this case we don't really have other use cases.